nvflare/private/fed/cmi.py

```diff
@@ -0,0 +1,252 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_constant import FLContextKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import ReservedHeaderKey, Shareable, make_reply
+from nvflare.fuel.f3.cellnet.cell import FQCN, Cell, Message, MessageHeaderKey
+from nvflare.fuel.f3.cellnet.cell import ReturnCode as CellReturnCode
+from nvflare.fuel.f3.cellnet.cell import new_message
+from nvflare.private.defs import CellMessageHeaderKeys
+
+
+class CellMessageInterface(FLComponent):
+
+    HEADER_KEY_PEER_PROPS = "cmi.peer_props"
+    HEADER_JOB_ID = "cmi.job_id"
+    HEADER_PROJECT_NAME = "cmi.project"
+    HEADER_SSID = "cmi.ssid"
+    HEADER_CLIENT_TOKEN = "cmi.client_token"
+    HEADER_CLIENT_NAME = "cmi.client_name"
+
+    PROP_KEY_CLIENT = "cmi.client"
+    PROP_KEY_FL_CTX = "cmi.fl_ctx"
+    PROP_KEY_PEER_CTX = "cmi.peer_ctx"
+
+    RC_TABLE = {
+        CellReturnCode.TIMEOUT: ReturnCode.COMMUNICATION_ERROR,
+        CellReturnCode.COMM_ERROR: ReturnCode.COMMUNICATION_ERROR,
+        CellReturnCode.PROCESS_EXCEPTION: ReturnCode.EXECUTION_EXCEPTION,
+        CellReturnCode.ABORT_RUN: CellReturnCode.ABORT_RUN,
+        CellReturnCode.INVALID_REQUEST: CellReturnCode.INVALID_REQUEST,
+        CellReturnCode.INVALID_SESSION: CellReturnCode.INVALID_SESSION,
+        CellReturnCode.AUTHENTICATION_ERROR: CellReturnCode.UNAUTHENTICATED,
+        CellReturnCode.SERVICE_UNAVAILABLE: CellReturnCode.SERVICE_UNAVAILABLE,
+    }
+
+    def __init__(
+        self,
+        engine,
+    ):
+        FLComponent.__init__(self)
+        self.engine = engine
+        self.cell = engine.get_cell()
+        self.ready = False
+
+        self.cell.add_incoming_request_filter(channel="*", topic="*", cb=self._filter_incoming_request)
+
+        self.cell.add_outgoing_reply_filter(channel="*", topic="*", cb=self._filter_outgoing_message)
+
+        self.cell.add_outgoing_request_filter(channel="*", topic="*", cb=self._filter_outgoing_message)
+
+        self.cell.add_incoming_reply_filter(channel="*", topic="*", cb=self._filter_incoming_message)
+
+    def new_cmi_message(self, fl_ctx: FLContext, headers=None, payload=None):
+        msg = new_message(headers, payload)
+        msg.set_prop(self.PROP_KEY_FL_CTX, fl_ctx)
+        return msg
+
+    def _filter_incoming_message(self, message: Message):
+        public_props = message.get_header(self.HEADER_KEY_PEER_PROPS)
+        if public_props:
+            peer_ctx = self._make_peer_ctx(public_props)
+            message.set_prop(self.PROP_KEY_PEER_CTX, peer_ctx)
+        shareable = message.payload
+        if isinstance(shareable, Shareable):
+            if public_props:
+                shareable.set_peer_props(public_props)
```

Comment:
It feels weird to modify the message payload if you are a messenger.

Reply:
`public_props` are special message headers in the message exchange. We are not modifying any other message payload.